File: maid-smart-mon.py
#!/usr/bin/env python3
"""
MAID SMART Health Monitoring System
Designed for Massive Array of Idle Disks environments where disk spin-ups are expensive.
Only monitors currently mounted/active disks to avoid unnecessary spin-ups.
"""

import sqlite3
import subprocess
import json
import re
import time
import logging
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Optional, Tuple
import argparse
import psutil

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('maid_smart_monitor.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

class MAIDSmartMonitor:
    """SMART monitoring system optimized for MAID environments"""
    
    # Target SMART attributes for monitoring
    TARGET_ATTRIBUTES = {
        1: 'Raw_Read_Error_Rate',
        3: 'Spin_Up_Time', 
        4: 'Start_Stop_Count',
        5: 'Reallocated_Sector_Ct',
        7: 'Seek_Error_Rate',
        9: 'Power_On_Hours',
        12: 'Power_Cycle_Count',
        187: 'Reported_Uncorrectable_Errors',
        188: 'Command_Timeout',
        190: 'Airflow_Temperature_Cel',
        191: 'G_Sense_Error_Rate',
        192: 'Power_Off_Retract_Count',
        193: 'Load_Cycle_Count',
        194: 'Temperature_Celsius',
        196: 'Reallocation_Event_Count',
        197: 'Current_Pending_Sector',
        198: 'Offline_Uncorrectable',
        199: 'UDMA_CRC_Error_Count',
        222: 'Loaded_Hours',
        240: 'Head_Flying_Hours',
        241: 'Total_LBAs_Written',
        242: 'Total_LBAs_Read'
    }
    
    def __init__(self, db_path: str = "maid_smart_data.db"):
        self.db_path = db_path
        self.init_database()
        
    def init_database(self):
        """Initialize SQLite database with required tables"""
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.cursor()
            
            # Main SMART data table
            cursor.execute('''
                CREATE TABLE IF NOT EXISTS smart_data (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    device TEXT NOT NULL,
                    serial_number TEXT,
                    model TEXT,
                    timestamp DATETIME NOT NULL,
                    attribute_id INTEGER NOT NULL,
                    attribute_name TEXT NOT NULL,
                    raw_value INTEGER,
                    normalized_value INTEGER,
                    threshold INTEGER,
                    worst_value INTEGER,
                    flags TEXT,
                    UNIQUE(device, timestamp, attribute_id)
                )
            ''')
            
            # Device status table to track spin state
            cursor.execute('''
                CREATE TABLE IF NOT EXISTS device_status (
                    device TEXT PRIMARY KEY,
                    serial_number TEXT,
                    model TEXT,
                    last_seen DATETIME,
                    is_mounted BOOLEAN,
                    mount_point TEXT,
                    smart_enabled BOOLEAN,
                    last_smart_check DATETIME,
                    spin_up_count INTEGER DEFAULT 0
                )
            ''')
            
            # Health alerts table
            cursor.execute('''
                CREATE TABLE IF NOT EXISTS health_alerts (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    device TEXT NOT NULL,
                    attribute_name TEXT NOT NULL,
                    alert_type TEXT NOT NULL,
                    message TEXT NOT NULL,
                    timestamp DATETIME NOT NULL,
                    resolved BOOLEAN DEFAULT FALSE
                )
            ''')
            
            conn.commit()
            logger.info(f"Database initialized: {self.db_path}")
    
    def get_mounted_drives(self) -> List[str]:
        """Get list of currently mounted drives to avoid spinning up idle disks"""
        mounted_drives = []
        
        # Get mounted partitions
        partitions = psutil.disk_partitions()
        
        for partition in partitions:
            # Extract device name (e.g., /dev/sda1 -> /dev/sda)
            device = re.sub(r'\d+$', '', partition.device)
            if device not in mounted_drives and device.startswith('/dev/sd'):
                mounted_drives.append(device)
        
        logger.info(f"Found {len(mounted_drives)} mounted drives: {mounted_drives}")
        return mounted_drives
    
    def check_smart_support(self, device: str) -> bool:
        """Check if device supports SMART without spinning it up"""
        try:
            # Use --nocheck=standby to avoid spinning up idle drives
            result = subprocess.run([
                'smartctl', '--nocheck=standby', '-i', device
            ], capture_output=True, text=True, timeout=30)
            
            if result.returncode == 0:
                return 'SMART support is: Enabled' in result.stdout
            else:
                logger.warning(f"SMART support check failed for {device}: {result.stderr}")
                return False
                
        except subprocess.TimeoutExpired:
            logger.error(f"Timeout checking SMART support for {device}")
            return False
        except Exception as e:
            logger.error(f"Error checking SMART support for {device}: {e}")
            return False
    
    def get_device_info(self, device: str) -> Tuple[Optional[str], Optional[str]]:
        """Get device serial number and model without spinning up"""
        try:
            result = subprocess.run([
                'smartctl', '--nocheck=standby', '-i', device
            ], capture_output=True, text=True, timeout=30)
            
            if result.returncode == 0:
                serial = None
                model = None
                
                for line in result.stdout.split('\n'):
                    if 'Serial Number:' in line:
                        serial = line.split('Serial Number:')[1].strip()
                    elif 'Device Model:' in line or 'Model Number:' in line:
                        model = line.split(':')[1].strip()
                
                return serial, model
            
        except Exception as e:
            logger.error(f"Error getting device info for {device}: {e}")
        
        return None, None
    
    def collect_smart_data(self, device: str) -> Dict:
        """Collect SMART data from a device (only if already spinning)"""
        try:
            # First check if device is in standby mode
            standby_check = subprocess.run([
                'smartctl', '--nocheck=standby', '-n', 'standby', device
            ], capture_output=True, text=True, timeout=10)
            
            if 'STANDBY' in standby_check.stdout:
                logger.info(f"Device {device} is in standby mode - skipping to avoid spin-up")
                return {}
            
            # Device is already spinning, safe to collect SMART data
            result = subprocess.run([
                'smartctl', '-A', '--json', device
            ], capture_output=True, text=True, timeout=60)
            
            if result.returncode == 0:
                data = json.loads(result.stdout)
                return data
            else:
                logger.warning(f"SMART data collection failed for {device}: {result.stderr}")
                return {}
                
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse SMART JSON for {device}: {e}")
            return {}
        except subprocess.TimeoutExpired:
            logger.error(f"Timeout collecting SMART data for {device}")
            return {}
        except Exception as e:
            logger.error(f"Error collecting SMART data for {device}: {e}")
            return {}
    
    def parse_smart_attributes(self, smart_data: Dict, device: str) -> List[Dict]:
        """Parse and filter SMART attributes for target IDs"""
        attributes = []
        
        if 'ata_smart_attributes' not in smart_data:
            return attributes
        
        table = smart_data['ata_smart_attributes'].get('table', [])
        
        for attr in table:
            attr_id = attr.get('id')
            if attr_id in self.TARGET_ATTRIBUTES:
                attributes.append({
                    'device': device,
                    'attribute_id': attr_id,
                    'attribute_name': self.TARGET_ATTRIBUTES[attr_id],
                    'raw_value': attr.get('raw', {}).get('value', 0),
                    'normalized_value': attr.get('value', 0),
                    'threshold': attr.get('thresh', 0),
                    'worst_value': attr.get('worst', 0),
                    'flags': str(attr.get('flags', {}))
                })
        
        return attributes
    
    def store_smart_data(self, attributes: List[Dict], serial: str, model: str):
        """Store SMART attributes in database"""
        if not attributes:
            return
        
        timestamp = datetime.now()
        
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.cursor()
            
            for attr in attributes:
                cursor.execute('''
                    INSERT OR REPLACE INTO smart_data 
                    (device, serial_number, model, timestamp, attribute_id, attribute_name,
                     raw_value, normalized_value, threshold, worst_value, flags)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ''', (
                    attr['device'], serial, model, timestamp,
                    attr['attribute_id'], attr['attribute_name'],
                    attr['raw_value'], attr['normalized_value'], 
                    attr['threshold'], attr['worst_value'], attr['flags']
                ))
            
            conn.commit()
            logger.info(f"Stored {len(attributes)} SMART attributes for {attributes[0]['device']}")
    
    def update_device_status(self, device: str, serial: str, model: str, 
                           is_mounted: bool, smart_enabled: bool):
        """Update device status in database"""
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.cursor()
            
            cursor.execute('''
                INSERT OR REPLACE INTO device_status
                (device, serial_number, model, last_seen, is_mounted, 
                 smart_enabled, last_smart_check)
                VALUES (?, ?, ?, ?, ?, ?, ?)
            ''', (device, serial, model, datetime.now(), is_mounted, 
                  smart_enabled, datetime.now()))
            
            conn.commit()
    
    def check_health_thresholds(self, attributes: List[Dict]):
        """Check for potential health issues and generate alerts"""
        critical_attrs = [5, 187, 196, 197, 198]  # Critical health indicators
        
        for attr in attributes:
            attr_id = attr['attribute_id']
            device = attr['device']
            raw_value = attr['raw_value']
            normalized_value = attr['normalized_value']
            threshold = attr['threshold']
            
            # Check for threshold violations
            if threshold > 0 and normalized_value <= threshold:
                self.create_alert(device, attr['attribute_name'], 'THRESHOLD_VIOLATION',
                                f"Value {normalized_value} below threshold {threshold}")
            
            # Check critical attributes
            if attr_id in critical_attrs and raw_value > 0:
                self.create_alert(device, attr['attribute_name'], 'CRITICAL_VALUE',
                                f"Non-zero critical value: {raw_value}")
            
            # Temperature warnings
            if attr_id in [190, 194] and raw_value > 60:  # Temperature > 60°C
                self.create_alert(device, attr['attribute_name'], 'HIGH_TEMPERATURE',
                                f"High temperature: {raw_value}°C")
    
    def create_alert(self, device: str, attribute: str, alert_type: str, message: str):
        """Create health alert in database"""
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.cursor()
            
            cursor.execute('''
                INSERT INTO health_alerts 
                (device, attribute_name, alert_type, message, timestamp)
                VALUES (?, ?, ?, ?, ?)
            ''', (device, attribute, alert_type, message, datetime.now()))
            
            conn.commit()
            logger.warning(f"HEALTH ALERT - {device}: {attribute} - {message}")
    
    def run_monitoring_cycle(self):
        """Run a single monitoring cycle"""
        logger.info("Starting SMART monitoring cycle...")
        
        mounted_drives = self.get_mounted_drives()
        
        for device in mounted_drives:
            try:
                logger.info(f"Processing device: {device}")
                
                # Get device info without spinning up
                serial, model = self.get_device_info(device)
                smart_enabled = self.check_smart_support(device)
                
                # Update device status
                self.update_device_status(device, serial, model, True, smart_enabled)
                
                if not smart_enabled:
                    logger.warning(f"SMART not supported/enabled on {device}")
                    continue
                
                # Collect SMART data (only if device is already spinning)
                smart_data = self.collect_smart_data(device)
                
                if smart_data:
                    attributes = self.parse_smart_attributes(smart_data, device)
                    if attributes:
                        self.store_smart_data(attributes, serial, model)
                        self.check_health_thresholds(attributes)
                    else:
                        logger.info(f"No target SMART attributes found for {device}")
                else:
                    logger.info(f"No SMART data collected for {device} (likely in standby)")
                    
            except Exception as e:
                logger.error(f"Error processing device {device}: {e}")
        
        logger.info("Monitoring cycle completed")
    
    def get_health_summary(self) -> Dict:
        """Get health summary from database"""
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.cursor()
            
            # Get latest data for each device
            cursor.execute('''
                SELECT device, COUNT(*) as alert_count
                FROM health_alerts 
                WHERE resolved = FALSE 
                GROUP BY device
            ''')
            
            alerts = dict(cursor.fetchall())
            
            # Get device count
            cursor.execute('SELECT COUNT(DISTINCT device) FROM device_status')
            device_count = cursor.fetchone()[0]
            
            return {
                'total_devices': device_count,
                'devices_with_alerts': len(alerts),
                'alerts_by_device': alerts
            }
    
    def export_data(self, output_file: str, days: int = 30):
        """Export SMART data to CSV for analysis"""
        import csv
        
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.cursor()
            
            cursor.execute('''
                SELECT * FROM smart_data 
                WHERE timestamp >= datetime('now', '-{} days')
                ORDER BY device, timestamp, attribute_id
            '''.format(days))
            
            with open(output_file, 'w', newline='') as csvfile:
                writer = csv.writer(csvfile)
                writer.writerow([desc[0] for desc in cursor.description])
                writer.writerows(cursor.fetchall())
        
        logger.info(f"Data exported to {output_file}")

def main():
    parser = argparse.ArgumentParser(description='MAID SMART Health Monitor')
    parser.add_argument('--db', default='maid_smart_data.db', 
                       help='Database file path')
    parser.add_argument('--interval', type=int, default=300,
                       help='Monitoring interval in seconds (default: 300)')
    parser.add_argument('--daemon', action='store_true',
                       help='Run as daemon')
    parser.add_argument('--export', type=str,
                       help='Export data to CSV file')
    parser.add_argument('--summary', action='store_true',
                       help='Show health summary')
    
    args = parser.parse_args()
    
    monitor = MAIDSmartMonitor(args.db)
    
    if args.export:
        monitor.export_data(args.export)
        return
    
    if args.summary:
        summary = monitor.get_health_summary()
        print(f"MAID SMART Health Summary:")
        print(f"Total devices: {summary['total_devices']}")
        print(f"Devices with alerts: {summary['devices_with_alerts']}")
        for device, count in summary['alerts_by_device'].items():
            print(f"  {device}: {count} alerts")
        return
    
    if args.daemon:
        logger.info(f"Starting MAID SMART monitor daemon (interval: {args.interval}s)")
        try:
            while True:
                monitor.run_monitoring_cycle()
                time.sleep(args.interval)
        except KeyboardInterrupt:
            logger.info("Daemon stopped by user")
    else:
        monitor.run_monitoring_cycle()

if __name__ == '__main__':
    main()

File: maid-smart-mon.go
package main

import (
	"database/sql"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"os/signal"
	"regexp"
	"strconv"
	"strings"
	"syscall"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

// SmartAttribute represents a SMART attribute from smartctl
type SmartAttribute struct {
	ID     int                    `json:"id"`
	Name   string                 `json:"name"`
	Value  int                    `json:"value"`
	Worst  int                    `json:"worst"`
	Thresh int                    `json:"thresh"`
	Raw    map[string]interface{} `json:"raw"`
	Flags  map[string]bool        `json:"flags"`
}

// SmartData represents the JSON output from smartctl
type SmartData struct {
	ATASmartAttributes struct {
		Table []SmartAttribute `json:"table"`
	} `json:"ata_smart_attributes"`
	ModelName    string `json:"model_name"`
	SerialNumber string `json:"serial_number"`
}

// DeviceInfo holds basic device information
type DeviceInfo struct {
	Device       string
	SerialNumber string
	Model        string
	IsMounted    bool
	SmartEnabled bool
}

// HealthAlert represents a health alert
type HealthAlert struct {
	Device        string
	AttributeName string
	AlertType     string
	Message       string
	Timestamp     time.Time
}

// MAIDSmartMonitor is the main monitoring system
type MAIDSmartMonitor struct {
	db            *sql.DB
	dbPath        string
	targetAttribs map[int]string
	logger        *log.Logger
}

// NewMAIDSmartMonitor creates a new monitor instance
func NewMAIDSmartMonitor(dbPath string) (*MAIDSmartMonitor, error) {
	// Target SMART attributes for monitoring
	targetAttribs := map[int]string{
		1:   "Raw_Read_Error_Rate",
		3:   "Spin_Up_Time",
		4:   "Start_Stop_Count",
		5:   "Reallocated_Sector_Ct",
		7:   "Seek_Error_Rate",
		9:   "Power_On_Hours",
		12:  "Power_Cycle_Count",
		187: "Reported_Uncorrectable_Errors",
		188: "Command_Timeout",
		190: "Airflow_Temperature_Cel",
		191: "G_Sense_Error_Rate",
		192: "Power_Off_Retract_Count",
		193: "Load_Cycle_Count",
		194: "Temperature_Celsius",
		196: "Reallocation_Event_Count",
		197: "Current_Pending_Sector",
		198: "Offline_Uncorrectable",
		199: "UDMA_CRC_Error_Count",
		222: "Loaded_Hours",
		240: "Head_Flying_Hours",
		241: "Total_LBAs_Written",
		242: "Total_LBAs_Read",
	}

	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %v", err)
	}

	monitor := &MAIDSmartMonitor{
		db:            db,
		dbPath:        dbPath,
		targetAttribs: targetAttribs,
		logger:        log.New(os.Stdout, "[MAID-SMART] ", log.LstdFlags),
	}

	if err := monitor.initDatabase(); err != nil {
		return nil, fmt.Errorf("failed to initialize database: %v", err)
	}

	return monitor, nil
}

// Close closes the database connection
func (m *MAIDSmartMonitor) Close() error {
	return m.db.Close()
}

// initDatabase initializes the SQLite database with required tables
func (m *MAIDSmartMonitor) initDatabase() error {
	queries := []string{
		`CREATE TABLE IF NOT EXISTS smart_data (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			device TEXT NOT NULL,
			serial_number TEXT,
			model TEXT,
			timestamp DATETIME NOT NULL,
			attribute_id INTEGER NOT NULL,
			attribute_name TEXT NOT NULL,
			raw_value INTEGER,
			normalized_value INTEGER,
			threshold INTEGER,
			worst_value INTEGER,
			flags TEXT,
			UNIQUE(device, timestamp, attribute_id)
		)`,
		`CREATE TABLE IF NOT EXISTS device_status (
			device TEXT PRIMARY KEY,
			serial_number TEXT,
			model TEXT,
			last_seen DATETIME,
			is_mounted BOOLEAN,
			mount_point TEXT,
			smart_enabled BOOLEAN,
			last_smart_check DATETIME,
			spin_up_count INTEGER DEFAULT 0
		)`,
		`CREATE TABLE IF NOT EXISTS health_alerts (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			device TEXT NOT NULL,
			attribute_name TEXT NOT NULL,
			alert_type TEXT NOT NULL,
			message TEXT NOT NULL,
			timestamp DATETIME NOT NULL,
			resolved BOOLEAN DEFAULT FALSE
		)`,
	}

	for _, query := range queries {
		if _, err := m.db.Exec(query); err != nil {
			return fmt.Errorf("failed to execute query: %v", err)
		}
	}

	m.logger.Printf("Database initialized: %s", m.dbPath)
	return nil
}

// getMountedDrives returns list of currently mounted drives to avoid spinning up idle disks
func (m *MAIDSmartMonitor) getMountedDrives() ([]string, error) {
	content, err := ioutil.ReadFile("/proc/mounts")
	if err != nil {
		return nil, fmt.Errorf("failed to read /proc/mounts: %v", err)
	}

	var mountedDrives []string
	deviceMap := make(map[string]bool)

	// Regex to match device names like /dev/sda1, /dev/nvme0n1p1, etc.
	deviceRegex := regexp.MustCompile(`^(/dev/[a-z]+)`)

	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		fields := strings.Fields(line)
		if len(fields) >= 2 {
			device := fields[0]
			if strings.HasPrefix(device, "/dev/sd") || strings.HasPrefix(device, "/dev/nvme") {
				// Extract base device name (e.g., /dev/sda1 -> /dev/sda)
				matches := deviceRegex.FindStringSubmatch(device)
				if len(matches) > 1 {
					baseDevice := matches[1]
					// Remove partition numbers for SATA drives
					baseDeviceClean := regexp.MustCompile(`\d+$`).ReplaceAllString(baseDevice, "")
					if !deviceMap[baseDeviceClean] {
						deviceMap[baseDeviceClean] = true
						mountedDrives = append(mountedDrives, baseDeviceClean)
					}
				}
			}
		}
	}

	m.logger.Printf("Found %d mounted drives: %v", len(mountedDrives), mountedDrives)
	return mountedDrives, nil
}

// checkSmartSupport checks if device supports SMART without spinning it up
func (m *MAIDSmartMonitor) checkSmartSupport(device string) bool {
	cmd := exec.Command("smartctl", "--nocheck=standby", "-i", device)
	output, err := cmd.Output()
	if err != nil {
		m.logger.Printf("SMART support check failed for %s: %v", device, err)
		return false
	}

	return strings.Contains(string(output), "SMART support is: Enabled")
}

// getDeviceInfo gets device serial number and model without spinning up
func (m *MAIDSmartMonitor) getDeviceInfo(device string) (string, string, error) {
	cmd := exec.Command("smartctl", "--nocheck=standby", "-i", device)
	output, err := cmd.Output()
	if err != nil {
		return "", "", fmt.Errorf("failed to get device info: %v", err)
	}

	var serial, model string
	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if strings.Contains(line, "Serial Number:") {
			parts := strings.Split(line, "Serial Number:")
			if len(parts) > 1 {
				serial = strings.TrimSpace(parts[1])
			}
		} else if strings.Contains(line, "Device Model:") || strings.Contains(line, "Model Number:") {
			parts := strings.Split(line, ":")
			if len(parts) > 1 {
				model = strings.TrimSpace(parts[1])
			}
		}
	}

	return serial, model, nil
}

// isDeviceInStandby checks if device is in standby mode
func (m *MAIDSmartMonitor) isDeviceInStandby(device string) bool {
	cmd := exec.Command("smartctl", "--nocheck=standby", "-n", "standby", device)
	output, err := cmd.Output()
	if err != nil {
		return false
	}

	return strings.Contains(string(output), "STANDBY")
}

// collectSmartData collects SMART data from a device (only if already spinning)
func (m *MAIDSmartMonitor) collectSmartData(device string) (*SmartData, error) {
	// First check if device is in standby mode
	if m.isDeviceInStandby(device) {
		m.logger.Printf("Device %s is in standby mode - skipping to avoid spin-up", device)
		return nil, nil
	}

	// Device is already spinning, safe to collect SMART data
	cmd := exec.Command("smartctl", "-A", "--json", device)
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to collect SMART data: %v", err)
	}

	var smartData SmartData
	if err := json.Unmarshal(output, &smartData); err != nil {
		return nil, fmt.Errorf("failed to parse SMART JSON: %v", err)
	}

	return &smartData, nil
}

// parseSmartAttributes parses and filters SMART attributes for target IDs
func (m *MAIDSmartMonitor) parseSmartAttributes(smartData *SmartData, device string) []map[string]interface{} {
	var attributes []map[string]interface{}

	for _, attr := range smartData.ATASmartAttributes.Table {
		if name, exists := m.targetAttribs[attr.ID]; exists {
			rawValue := int64(0)
			if val, ok := attr.Raw["value"]; ok {
				switch v := val.(type) {
				case float64:
					rawValue = int64(v)
				case int64:
					rawValue = v
				case string:
					if parsed, err := strconv.ParseInt(v, 10, 64); err == nil {
						rawValue = parsed
					}
				}
			}

			attributes = append(attributes, map[string]interface{}{
				"device":           device,
				"attribute_id":     attr.ID,
				"attribute_name":   name,
				"raw_value":        rawValue,
				"normalized_value": attr.Value,
				"threshold":        attr.Thresh,
				"worst_value":      attr.Worst,
				"flags":            fmt.Sprintf("%+v", attr.Flags),
			})
		}
	}

	return attributes
}

// storeSmartData stores SMART attributes in database
func (m *MAIDSmartMonitor) storeSmartData(attributes []map[string]interface{}, serial, model string) error {
	if len(attributes) == 0 {
		return nil
	}

	timestamp := time.Now()

	tx, err := m.db.Begin()
	if err != nil {
		return fmt.Errorf("failed to begin transaction: %v", err)
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`
		INSERT OR REPLACE INTO smart_data 
		(device, serial_number, model, timestamp, attribute_id, attribute_name,
		 raw_value, normalized_value, threshold, worst_value, flags)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`)
	if err != nil {
		return fmt.Errorf("failed to prepare statement: %v", err)
	}
	defer stmt.Close()

	for _, attr := range attributes {
		_, err := stmt.Exec(
			attr["device"], serial, model, timestamp,
			attr["attribute_id"], attr["attribute_name"],
			attr["raw_value"], attr["normalized_value"],
			attr["threshold"], attr["worst_value"], attr["flags"],
		)
		if err != nil {
			return fmt.Errorf("failed to insert attribute: %v", err)
		}
	}

	if err := tx.Commit(); err != nil {
		return fmt.Errorf("failed to commit transaction: %v", err)
	}

	m.logger.Printf("Stored %d SMART attributes for %s", len(attributes), attributes[0]["device"])
	return nil
}

// updateDeviceStatus updates device status in database
func (m *MAIDSmartMonitor) updateDeviceStatus(device, serial, model string, isMounted, smartEnabled bool) error {
	_, err := m.db.Exec(`
		INSERT OR REPLACE INTO device_status
		(device, serial_number, model, last_seen, is_mounted, 
		 smart_enabled, last_smart_check)
		VALUES (?, ?, ?, ?, ?, ?, ?)
	`, device, serial, model, time.Now(), isMounted, smartEnabled, time.Now())

	return err
}

// checkHealthThresholds checks for potential health issues and generates alerts
func (m *MAIDSmartMonitor) checkHealthThresholds(attributes []map[string]interface{}) {
	criticalAttrs := map[int]bool{5: true, 187: true, 196: true, 197: true, 198: true}

	for _, attr := range attributes {
		attrID := attr["attribute_id"].(int)
		device := attr["device"].(string)
		rawValue := attr["raw_value"].(int64)
		normalizedValue := attr["normalized_value"].(int)
		threshold := attr["threshold"].(int)
		attrName := attr["attribute_name"].(string)

		// Check for threshold violations
		if threshold > 0 && normalizedValue <= threshold {
			m.createAlert(device, attrName, "THRESHOLD_VIOLATION",
				fmt.Sprintf("Value %d below threshold %d", normalizedValue, threshold))
		}

		// Check critical attributes
		if criticalAttrs[attrID] && rawValue > 0 {
			m.createAlert(device, attrName, "CRITICAL_VALUE",
				fmt.Sprintf("Non-zero critical value: %d", rawValue))
		}

		// Temperature warnings
		if (attrID == 190 || attrID == 194) && rawValue > 60 {
			m.createAlert(device, attrName, "HIGH_TEMPERATURE",
				fmt.Sprintf("High temperature: %d°C", rawValue))
		}
	}
}

// createAlert creates health alert in database
func (m *MAIDSmartMonitor) createAlert(device, attribute, alertType, message string) {
	_, err := m.db.Exec(`
		INSERT INTO health_alerts 
		(device, attribute_name, alert_type, message, timestamp)
		VALUES (?, ?, ?, ?, ?)
	`, device, attribute, alertType, message, time.Now())

	if err != nil {
		m.logger.Printf("Failed to create alert: %v", err)
	} else {
		m.logger.Printf("HEALTH ALERT - %s: %s - %s", device, attribute, message)
	}
}

// runMonitoringCycle runs a single monitoring cycle
func (m *MAIDSmartMonitor) runMonitoringCycle() error {
	m.logger.Println("Starting SMART monitoring cycle...")

	mountedDrives, err := m.getMountedDrives()
	if err != nil {
		return fmt.Errorf("failed to get mounted drives: %v", err)
	}

	for _, device := range mountedDrives {
		m.logger.Printf("Processing device: %s", device)

		// Get device info without spinning up
		serial, model, err := m.getDeviceInfo(device)
		if err != nil {
			m.logger.Printf("Failed to get device info for %s: %v", device, err)
			continue
		}

		smartEnabled := m.checkSmartSupport(device)

		// Update device status
		if err := m.updateDeviceStatus(device, serial, model, true, smartEnabled); err != nil {
			m.logger.Printf("Failed to update device status for %s: %v", device, err)
		}

		if !smartEnabled {
			m.logger.Printf("SMART not supported/enabled on %s", device)
			continue
		}

		// Collect SMART data (only if device is already spinning)
		smartData, err := m.collectSmartData(device)
		if err != nil {
			m.logger.Printf("Error collecting SMART data for %s: %v", device, err)
			continue
		}

		if smartData != nil {
			attributes := m.parseSmartAttributes(smartData, device)
			if len(attributes) > 0 {
				if err := m.storeSmartData(attributes, serial, model); err != nil {
					m.logger.Printf("Failed to store SMART data for %s: %v", device, err)
				} else {
					m.checkHealthThresholds(attributes)
				}
			} else {
				m.logger.Printf("No target SMART attributes found for %s", device)
			}
		} else {
			m.logger.Printf("No SMART data collected for %s (likely in standby)", device)
		}
	}

	m.logger.Println("Monitoring cycle completed")
	return nil
}

// getHealthSummary gets health summary from database
func (m *MAIDSmartMonitor) getHealthSummary() (map[string]interface{}, error) {
	// Get alerts by device
	rows, err := m.db.Query(`
		SELECT device, COUNT(*) as alert_count
		FROM health_alerts 
		WHERE resolved = FALSE 
		GROUP BY device
	`)
	if err != nil {
		return nil, fmt.Errorf("failed to query alerts: %v", err)
	}
	defer rows.Close()

	alertsByDevice := make(map[string]int)
	for rows.Next() {
		var device string
		var count int
		if err := rows.Scan(&device, &count); err != nil {
			return nil, fmt.Errorf("failed to scan alert row: %v", err)
		}
		alertsByDevice[device] = count
	}

	// Get device count
	var deviceCount int
	err = m.db.QueryRow("SELECT COUNT(DISTINCT device) FROM device_status").Scan(&deviceCount)
	if err != nil {
		return nil, fmt.Errorf("failed to get device count: %v", err)
	}

	return map[string]interface{}{
		"total_devices":       deviceCount,
		"devices_with_alerts": len(alertsByDevice),
		"alerts_by_device":    alertsByDevice,
	}, nil
}

// exportData exports SMART data to CSV for analysis
func (m *MAIDSmartMonitor) exportData(outputFile string, days int) error {
	rows, err := m.db.Query(`
		SELECT * FROM smart_data 
		WHERE timestamp >= datetime('now', '-' || ? || ' days')
		ORDER BY device, timestamp, attribute_id
	`, days)
	if err != nil {
		return fmt.Errorf("failed to query data: %v", err)
	}
	defer rows.Close()

	file, err := os.Create(outputFile)
	if err != nil {
		return fmt.Errorf("failed to create output file: %v", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write header
	columns, err := rows.Columns()
	if err != nil {
		return fmt.Errorf("failed to get columns: %v", err)
	}
	writer.Write(columns)

	// Write data
	for rows.Next() {
		values := make([]interface{}, len(columns))
		valuePtrs := make([]interface{}, len(columns))
		for i := range values {
			valuePtrs[i] = &values[i]
		}

		if err := rows.Scan(valuePtrs...); err != nil {
			return fmt.Errorf("failed to scan row: %v", err)
		}

		record := make([]string, len(columns))
		for i, val := range values {
			if val != nil {
				record[i] = fmt.Sprintf("%v", val)
			}
		}
		writer.Write(record)
	}

	m.logger.Printf("Data exported to %s", outputFile)
	return nil
}

func main() {
	var (
		dbPath   = flag.String("db", "maid_smart_data.db", "Database file path")
		interval = flag.Int("interval", 300, "Monitoring interval in seconds")
		daemon   = flag.Bool("daemon", false, "Run as daemon")
		export   = flag.String("export", "", "Export data to CSV file")
		summary  = flag.Bool("summary", false, "Show health summary")
	)
	flag.Parse()

	monitor, err := NewMAIDSmartMonitor(*dbPath)
	if err != nil {
		log.Fatalf("Failed to create monitor: %v", err)
	}
	defer monitor.Close()

	if *export != "" {
		if err := monitor.exportData(*export, 30); err != nil {
			log.Fatalf("Failed to export data: %v", err)
		}
		return
	}

	if *summary {
		summary, err := monitor.getHealthSummary()
		if err != nil {
			log.Fatalf("Failed to get health summary: %v", err)
		}

		fmt.Println("MAID SMART Health Summary:")
		fmt.Printf("Total devices: %v\n", summary["total_devices"])
		fmt.Printf("Devices with alerts: %v\n", summary["devices_with_alerts"])

		if alerts, ok := summary["alerts_by_device"].(map[string]int); ok {
			for device, count := range alerts {
				fmt.Printf("  %s: %d alerts\n", device, count)
			}
		}
		return
	}

	if *daemon {
		monitor.logger.Printf("Starting MAID SMART monitor daemon (interval: %ds)", *interval)

		// Set up signal handling for graceful shutdown
		sigChan := make(chan os.Signal, 1)
		signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

		ticker := time.NewTicker(time.Duration(*interval) * time.Second)
		defer ticker.Stop()

		// Run initial cycle
		if err := monitor.runMonitoringCycle(); err != nil {
			monitor.logger.Printf("Error in monitoring cycle: %v", err)
		}

		for {
			select {
			case <-ticker.C:
				if err := monitor.runMonitoringCycle(); err != nil {
					monitor.logger.Printf("Error in monitoring cycle: %v", err)
				}
			case sig := <-sigChan:
				monitor.logger.Printf("Received signal %v, shutting down...", sig)
				return
			}
		}
	} else {
		if err := monitor.runMonitoringCycle(); err != nil {
			log.Fatalf("Error in monitoring cycle: %v", err)
		}
	}
}
